feat(editor): Allow jumping into sub-workflow with shortkey

RicardoE105 opened the pull request:
## Summary
  
This PR adds a new context menu item when right-clicking a sub-workflow node, allowing users to quickly open sub-workflows in a new tab. This enhances workflow navigation and improves user experience when working with complex nested workflows.

### Features

- Adds "Open Sub-workflow" option to the context menu for sub-workflow nodes - Option only shows when the node is the execute node and will be disabled if there not workflow selected in the node.
- Implements keyboard shortcuts for faster access:
    - **Windows**: Ctrl + Shift + O or Ctrl + Double-click
    - **macOS**: ⌘ + Shift + O or ⌘ + Double-click

### Demo

https://share.cleanshot.com/vtBXn5HZ

## Related Linear tickets, Github issues, and Community forum posts

https://linear.app/n8n/issue/ADO-3422/feature-jump-into-sub-workflow-with-shortkey

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

RicardoE105 committed:
Jump into sub-workflow with shortkey

RicardoE105 committed:
remove unused imports

RicardoE105 committed:
keep old behavior

RicardoE105 committed:
Remove unnecessary changes

RicardoE105 committed:
add tests

RicardoE105 committed:
remove unnecessary return

RicardoE105 commented:
@CharlieKolb Adding a test for clicking the option in the context menu was straightforward. However, I couldn’t find any unit test examples for triggering the action via shortcuts like `Cmd + Shift + O` or `Cmd + double-click`. Maybe an E2E test is the best approach here—thoughts?

CharlieKolb commented on packages/frontend/editor-ui/src/stores/nodeTypes.store.ts:
```diff
@@ -328,6 +333,16 @@ export const useNodeTypesStore = defineStore(STORES.NODE_TYPES, () => {
 		return await nodeTypesApi.getNodeParameterActionResult(rootStore.restApiContext, sendData);
 	};
 
+	function getSubworkflowId(node: INodeUi): string | undefined {
```
Should this be in `nodeTypes.store.ts`? How do you feel about putting it in `NodeHelpers` instead?

RicardoE105 replied:
Good point. I was not sure where to add it, and completely forgot `NodeHelpers` exist. Definitely a better place. Added it there, but had to move some stuff around to break the dependency cycle that came from moving to the NodeHelpers.ts.

CharlieKolb replied:
Ah sorry for the trouble :D

CharlieKolb commented on packages/frontend/editor-ui/src/views/NodeView.vue:
```diff
@@ -666,10 +667,22 @@ function onClickNode() {
+function onSetNodeActivated(id: string, event?: MouseEvent) {
+	// Handle Ctrl/Cmd + Double Click case
+	if (event?.metaKey || event?.ctrlKey) {
+		const couldOpen = tryToOpenSubworkflowInNewTab(id);
```
[nit] `didOpen` or `succeeded` is more clear, `could` sounds like we only confirmed we're able to

CharlieKolb commented on packages/frontend/editor-ui/src/composables/useContextMenu.ts:
```diff
@@ -253,6 +269,15 @@ export const useContextMenu = (onAction: ContextMenuActionCallback = () => {}) =
+						id: 'open_sub_workflow',
+						label: i18n.baseText('contextMenu.openSubworkflow'),
+						shortcut: { shiftKey: true, metaKey: true, keys: ['O'] },
+						disabled: !canOpenSubworkflow.value || isReadOnly.value,
```
Any particular reason we disallow opening read-only workflows?

RicardoE105 replied:
Good question. I guess for this particular option it's not needed.

RicardoE105 replied:
Removed it.

CharlieKolb reviewed (commented):
Really nice work with the feature! I particularly like how you handle the double click implementation.
Left mostly nits and a question regarding disabling the option on read-only

I'm content with the tests as is tbh, your call whether you want an e2e test, or maybe one for `tryToOpenSubworkflowInNewTab`

[ux feedback] Having `Open...` and `Open Sub-workflow` is a bit unfortunate. If this wasn't discussed yet, maybe bring it up with Rob to use `Jump to` or `Go to` to mirror IDE language more?

Is supporting the workflow tool out of scope?

RicardoE105 committed:
rename `id` to `nodeId` in `tryToOpenSubworkflowInNewTab` method

RicardoE105 committed:
move `getSubworklowId`from nodeTypesStore to NodeHelpers

RicardoE105 committed:
rename `couldOpen` to `didOpen`

RicardoE105 committed:
do not disable open subworkflow option when workflow is read only

RicardoE105 committed:
Merge branch 'master' into ado-3422-feature-jump-into-sub-workflow-with-shortkey

RicardoE105 committed:
Remove `isResourceLocatorValue` type guard duplication

We had the same function in `editor-ui/src/utils` and `workflows/src/type-guard.ts`. Kept the one in workflows packages as this is package shared across the main packages.

RicardoE105 committed:
use `getSubworkflowId` from right import in `useContextMenu.ts`

RicardoE105 committed:
remove unused imports

CharlieKolb reviewed (approved):
Happy to do the workflow tool (and maybe name change) in a follow up 🚀
